File: legacy/Source Code (py)/Window/_subclass.py
from PyQt5.QtWidgets import QFrame, QSizePolicy, QLabel, QHBoxLayout, QDialog, QVBoxLayout, \
	QCheckBox, QLineEdit, QPushButton, QListWidget, QAbstractItemView
from PyQt5.QtCore import Qt


class QHSeperationLine(QFrame):
	'''
	a horizontal seperation line\n
	'''
	def __init__(self):
		super().__init__()
		# self.setMinimumWidth(1)
		self.setFixedHeight(2)
		self.setFrameShape(QFrame.HLine)
		self.setFrameShadow(QFrame.Sunken)
		self.setSizePolicy(QSizePolicy.Preferred, QSizePolicy.Minimum)

class ModQHBoxLayout(QHBoxLayout):
	def __init__(self):
		super().__init__()
		self.setAlignment(Qt.AlignTop)

class ModQLabel(QLabel):
	def __init__(self, str):
		super().__init__(str)

		self.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Maximum)
		self.setAlignment(Qt.AlignHCenter | Qt.AlignTop)

class MessageBoxwLabel(QDialog):
	folder_name = None

	def __init__(self, folder):
		super().__init__()

		self.mainlayout = QVBoxLayout()

		folder_label = QLabel(f"Folder: {folder}")
		self.mainlayout.addWidget(folder_label)

		self.addtag = QCheckBox("Don't add a tag")
		self.mainlayout.addWidget(self.addtag)

		label = QLabel("Custom Tag?")
		self.mainlayout.addWidget(label)

		self.lineedit = QLineEdit()
		self.mainlayout.addWidget(self.lineedit)

		self.button = QPushButton("Done")
		self.mainlayout.addWidget(self.button)
		self.button.clicked.connect(self.click)

		self.setLayout(self.mainlayout)
		
		self.returned = self.exec_()
	
	def click(self):
		print("click")

		self.tag_from_popup = self.lineedit.text().strip()
		self.addtag = self.addtag.isChecked()
		self.done(1)

class ListBoxWidget(QListWidget):
	dir_list = []

	def __init__(self):
		super().__init__()
		self.viewport().setAcceptDrops(True)
		self.setSizePolicy(QSizePolicy.Expanding,QSizePolicy.Expanding)
		self.setSelectionMode(QAbstractItemView.ExtendedSelection)
		self.setDragDropMode(QAbstractItemView.DragDrop)
		self.setDropIndicatorShown(True)

	def dragEnterEvent(self, event):
		# self.passed_m = event.mimeData()
		event.accept()

	def dragMoveEvent(self, event):
		# print("dragMoveEvent")
		if not event.mimeData().hasUrls():
			# print("no urls")
			event.ignore()
			return
		event.accept()

	def dropEvent(self, event):
		if event.mimeData().hasUrls():
			# event.setDropAction(Qt.CopyAction)
			event.accept()

			links = []
			for url in event.mimeData().urls():
				# https://doc.qt.io/qt-5/qurl.html
				if url.isLocalFile():
					links.append(str(url.toLocalFile()))
				else:
					links.append(str(url.toString()))
			self.addItems(links)
		else:
			event.ignore()
	
	def getAllItems(self):
		for item in range(self.count()):
			dir = self.item(item).text()
			self.dir_list.append(dir)

		ret = self.dir_list 
		self.dir_list = []
		return ret

File: src/Window/subclass.cpp
#include "../../Include/Window/subclass.h"

// + QHSeparationLine
QHSeparationLine::QHSeparationLine() {
    this->setFixedHeight(2);
    this->setFrameShape(QFrame::HLine);
    this->setFrameShadow(QFrame::Sunken);
    this->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Minimum);
}

// + ModQHBoxLayout
ModQHBoxLayout::ModQHBoxLayout() {
    this->setAlignment(Qt::AlignTop);
}

// + ModQLabel
ModQLabel::ModQLabel(QString str) : QLabel(str) {
    this->setSizePolicy(QSizePolicy::Maximum, QSizePolicy::Maximum);
    this->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
}

// + MessageBoxwLabel
MessageBoxwLabel::MessageBoxwLabel(QString folder) {
    QVBoxLayout* mainlayout = new QVBoxLayout();

    QLabel* folderLabel = new QLabel("Folder: " + folder);
    mainlayout->addWidget(folderLabel);

    addTag_chkbx = new QCheckBox("Don't add a tag");
    mainlayout->addWidget(addTag_chkbx);

    QLabel* label = new QLabel("Custom Tag?");
    mainlayout->addWidget(label);

    lineEdit = new QLineEdit();
    mainlayout->addWidget(lineEdit);

    QPushButton* button = new QPushButton("Done");
    QObject::connect(button, QPushButton::clicked, this, &click);
    mainlayout->addWidget(button);

    this->setLayout(mainlayout);

    int returned = this->exec();
}

void MessageBoxwLabel::click() {
    qDebug() << "click";

    tag_from_popup = this->lineEdit->text().simplified();  // lineedit.text();
    addTag = this->addTag_chkbx->isChecked();              // lineedit.text();
    this->done(1);
}

// + ListBoxwidget
ListBoxwidget::ListBoxwidget() {
    this->viewport()->setAcceptDrops(true);
    this->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    this->setSelectionMode(QAbstractItemView::ExtendedSelection);
    this->setDragDropMode(QAbstractItemView::DragDrop);
    this->setDropIndicatorShown(true);
}

void ListBoxwidget::dragEnterEvent(QDragEnterEvent* event) {
    event->accept();
}

void ListBoxwidget::dragMoveEvent(QDragMoveEvent* event) {
    if (!(event->mimeData()->hasUrls())) {
        event->ignore();
        return;
    }

    event->accept();
}

void ListBoxwidget::dropEvent(QDropEvent* event) {
    if (event->mimeData()->hasUrls()) {
        event->accept();

        QList<QString> tmp_list;

        for (QUrl url : event->mimeData()->urls()) {
            if (url.isLocalFile()) {
                tmp_list.append(url.toLocalFile());
            }
        }
        this->ov_addItems(tmp_list);
        qDebug() << "Dropping";
    } else {
        event->ignore();
    }
}

QList<QString> ListBoxwidget::getAllItems() {
    this->dir_list.clear();

    for (int i = 0; i != this->count(); i++) {
        QString url = this->item(i)->text();
        this->dir_list.append(url);
    }
    return this->dir_list;
}

QStringList ListBoxwidget::removeDuplicates(const QStringList labels) {
    QStringList nodup;

    QStringList contents = this->getAllItems();

    for (auto i : labels) {
        if (!contents.contains(i)) {
            nodup.append(i);
        }
    }

    return nodup;
}

void ListBoxwidget::ov_addItems(const QStringList labels) {
    // remove item on the list that are duplicates

    // (const_cast<QStringList*>(labels))->removeDuplicates();

    qDebug() << (labels);

    QStringList nodup = this->removeDuplicates(labels);

    qDebug() << (nodup);
    this->addItems(nodup);
}

// + QGroupBox
ModGroupBox::ModGroupBox(QString title) : QGroupBox(title) {
    this->setCheckable(true);
    this->setChecked(true);
}